### PR TITLE
Fix parser error recovery

### LINKED ISSUE #24890: IllegalStateException list-bp-or-list-cons-member when parsing invalid syntax
**Description:**
`list-bp-or-list-cons-member` context is not handled in parser error handler.
consider the following code
```
public function foo() {
    [{a}, #] = t;
}
```
Parser crashes with bellow stack trace
```
Exception in thread "main" java.lang.IllegalStateException: cannot find the next rule for: list-bp-or-list-cons-member
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.getNextRuleForKeywords(BallerinaParserErrorHandler.java:2737)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.getNextRuleInternal(BallerinaParserErrorHandler.java:2522)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.getNextRule(BallerinaParserErrorHandler.java:2284)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:998)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:197)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.seekMatch(AbstractParserErrorHandler.java:179)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.recover(AbstractParserErrorHandler.java:102)
	at io.ballerinalang.compiler.internal.parser.AbstractParser.recover(AbstractParser.java:85)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseListBindingPatternOrListConstructorMember(BallerinaParser.java:16355)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseListBindingPatternOrListConstructor(BallerinaParser.java:16300)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseAsListBindingPatternOrListConstructor(BallerinaParser.java:15434)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseStatementStartsWithOpenBracket(BallerinaParser.java:15079)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseStatementStartsWithOpenBracket(BallerinaParser.java:15039)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseStatement(BallerinaParser.java:3800)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseStatement(BallerinaParser.java:3717)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseStatement(BallerinaParser.java:3636)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFunctionBodyBlock(BallerinaParser.java:2540)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFunctionBody(BallerinaParser.java:2481)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFunctionBody(BallerinaParser.java:2466)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.createFuncDefOrMethodDecl(BallerinaParser.java:1509)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFunctionKeywordRhs(BallerinaParser.java:1496)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFunctionKeywordRhs(BallerinaParser.java:1466)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFuncDefOrFuncTypeDesc(BallerinaParser.java:1451)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:1329)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:809)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:738)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseCompUnit(BallerinaParser.java:638)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parse(BallerinaParser.java:111)
	at io.ballerinalang.compiler.parser.test.SyntaxTreeJSONGenerator.getParserTree(SyntaxTreeJSONGenerator.java:98)
	at io.ballerinalang.compiler.parser.test.SyntaxTreeJSONGenerator.generateJSON(SyntaxTreeJSONGenerator.java:92)
	at io.ballerinalang.compiler.parser.test.SyntaxTreeJSONGenerator.generateJSON(SyntaxTreeJSONGenerator.java:83)
	at io.ballerinalang.compiler.parser.test.SyntaxTreeJSONGenerator.main(SyntaxTreeJSONGenerator.java:76)
```
**Suggested Labels (optional):**

- Component/Parser
- Type/Bug

**Suggested Assignees (optional):**

### LINKED ISSUE #24891: IllegalStateException tuple-type-desc-or-list-cont-member when parsing invalid syntax
**Description:**
`tuple-type-desc-or-list-cont-member` context is not handled in parser error handler.
consider the following code
```
public function foo() {
     [a[5], #] = t;
}
```
Parser crashes with bellow stack trace
```
Exception in thread "main" java.lang.IllegalStateException: cannot find the next rule for: tuple-type-desc-or-list-cont-member
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.getNextRuleForKeywords(BallerinaParserErrorHandler.java:2737)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.getNextRuleInternal(BallerinaParserErrorHandler.java:2522)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.getNextRule(BallerinaParserErrorHandler.java:2284)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:998)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:197)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.seekMatch(AbstractParserErrorHandler.java:179)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.recover(AbstractParserErrorHandler.java:102)
	at io.ballerinalang.compiler.internal.parser.AbstractParser.recover(AbstractParser.java:85)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseTupleTypeDescOrListConstructorMember(BallerinaParser.java:15301)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseTupleTypeDescOrListConstructor(BallerinaParser.java:15222)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseAsTupleTypeDescOrListConstructor(BallerinaParser.java:15197)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseStatementStartsWithOpenBracket(BallerinaParser.java:15081)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseStatementStartsWithOpenBracket(BallerinaParser.java:15039)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseStatement(BallerinaParser.java:3800)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseStatement(BallerinaParser.java:3717)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseStatement(BallerinaParser.java:3636)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFunctionBodyBlock(BallerinaParser.java:2540)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFunctionBody(BallerinaParser.java:2481)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFunctionBody(BallerinaParser.java:2466)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.createFuncDefOrMethodDecl(BallerinaParser.java:1509)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFunctionKeywordRhs(BallerinaParser.java:1496)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFunctionKeywordRhs(BallerinaParser.java:1466)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseFuncDefOrFuncTypeDesc(BallerinaParser.java:1451)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:1329)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:809)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:738)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parseCompUnit(BallerinaParser.java:638)
	at io.ballerinalang.compiler.internal.parser.BallerinaParser.parse(BallerinaParser.java:111)
	at io.ballerinalang.compiler.parser.test.SyntaxTreeJSONGenerator.getParserTree(SyntaxTreeJSONGenerator.java:98)
	at io.ballerinalang.compiler.parser.test.SyntaxTreeJSONGenerator.generateJSON(SyntaxTreeJSONGenerator.java:92)
	at io.ballerinalang.compiler.parser.test.SyntaxTreeJSONGenerator.generateJSON(SyntaxTreeJSONGenerator.java:83)
	at io.ballerinalang.compiler.parser.test.SyntaxTreeJSONGenerator.main(SyntaxTreeJSONGenerator.java:76)
```
**Suggested Labels (optional):**

- Component/Parser
- Type/Bug

### PR DESCRIPTION
**Pull request by m36dot**
## Purpose
Fix error recovery as some rules are not implemented in the errorhandler hence throw illegal state exception.

Fixes #24891 
Fixes #24890 

## Approach
Add rules for tuple-type-desc-or-list-cont-member and
list-bp-or-list-cons-member contexts, also added missing positions in
BLangNodeTransformer for ListBindingPatternNode and TupleDestructureStatement.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -618,12 +618,15 @@ public STNode resumeParsing(ParserRuleContext context, Object... args) {
                 return parseArgsBindingPatternEnd();
             case TABLE_ROW_END:
                 return parseTableRowEnd();
+            case LIST_BP_OR_LIST_CONSTRUCTOR_MEMBER:
+                return parseListBindingPatternOrListConstructorMember(peek().kind);
+            case TUPLE_TYPE_DESC_OR_LIST_CONST_MEMBER:
+                return parseTupleTypeDescOrListConstructorMember(peek().kind, (STNode) args[0]);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:624]: we haven't used peek() inside the resumeParsing method. Let's stick to a single pattern.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -15671,7 +15674,8 @@ private STNode parseTupleTypeDescOrListConstructorMember(SyntaxKind nextTokenKin
                     return parseTypeDescriptor(ParserRuleContext.TYPE_DESC_IN_TUPLE);
                 }
 
-                Solution solution = recover(peek(), ParserRuleContext.TUPLE_TYPE_DESC_OR_LIST_CONST_MEMBER);
+                Solution solution = recover(peek(), ParserRuleContext.TUPLE_TYPE_DESC_OR_LIST_CONST_MEMBER,
+                        annots);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:15678]: Not related to your PR.. the method-called for re-parse the same rule after inserting a token seems wrong.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 1 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:624]: alright

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 2 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:15682]: fixed

**Pull request merged**